src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/GenerateMethodCodeActionProvider.cs

```diff
@@ -0,0 +1,133 @@
+        }
+
+        return s_emptyResult;
+    }
+
+    private static List<RazorVSInternalCodeAction> CreateCodeAction(GenerateMethodCodeActionParams @params)
+    {
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            Action = LanguageServerConstants.CodeActions.GenerateMethod,
+            Language = LanguageServerConstants.CodeActions.Languages.Razor,
+            Data = @params,
+        };
+
+        var codeAction = RazorCodeActionFactory.CreateGenerateMethod(resolutionParams);
+        return new List<RazorVSInternalCodeAction> { codeAction };
+    }
+
+    private static bool IsGenerateEventHandlerValid(SyntaxNode owner, RazorCodeActionContext context, IEnumerable<Diagnostic> nameNotExistDiagnostics, [NotNullWhen(true)] out GenerateMethodCodeActionParams? @params)
+    {
+        @params = null;
+
+        // The owner should have a SyntaxKind of CSharpExpressionLiteral or MarkupTextLiteral.
+        // MarkupTextLiteral if the cursor is directly before the first letter of the method name.
+        // CSharpExpressionalLiteral if cursor is anywhere else in the method name.
+        if (owner.Kind != SyntaxKind.CSharpExpressionLiteral && owner.Kind != SyntaxKind.MarkupTextLiteral)
+        {
+            return false;
+        }
+
+        var parent = owner.Kind == SyntaxKind.CSharpExpressionLiteral ? owner.Parent.Parent : owner.Parent;
+        if (parent.Kind != SyntaxKind.MarkupTagHelperDirectiveAttribute)
+        {
+            return false;
+        }
+
+        var methodName = string.Empty;
+        if (owner.Kind == SyntaxKind.CSharpExpressionLiteral)
+        {
+            if (!TryParseUndefinedMethodName(owner, nameNotExistDiagnostics, out methodName))
+            {
+                return false;
+            }
+        }
+        else
+        {
+            var children = parent.ChildNodes();
+            foreach (var child in children)
+            {
+                if (child.Kind == SyntaxKind.MarkupTagHelperAttributeValue && TryParseUndefinedMethodName(child, nameNotExistDiagnostics, out methodName))
+                {
+                    break;
+                }
+            }
+        }
+
+        if (methodName.IsNullOrEmpty())
+        {
+            return false;
+        }
+
+        @params = new GenerateMethodCodeActionParams()
+        {
+            Uri = context.Request.TextDocument.Uri,
+            MethodName = methodName,
+        };
+
+        return true;
+    }
+
+    private static bool TryParseUndefinedMethodName(SyntaxNode node, IEnumerable<Diagnostic> nameNotExistDiagnostics, out string? methodName)
+    {
+        methodName = null;
+        var content = node.GetContent();
+        if (!SyntaxFacts.IsValidIdentifier(content))
+        {
+            return false;
+        }
+
+        if (!nameNotExistDiagnostics.Any(d => d.Message == $"The name '{content}' does not exist in the current context"))
```

Comment:
Filtering on the message like this is not going to work, sadly, because the messages could be translated into other languages by Roslyn. It's also reasonably fragile, which is not great. What kinds of false positives do we get if we just look for the error code, and don't worry about the message?

Reply:
I could be wrong, but I thought the code action context only included diagnostics that were relevant at the position the code action was invoked. So if the caret is on `ThisMethodExists` and the user clicks the lightbulb, I would expect the context to not contain any diagnostics.

Reply:
Just checked again and you are right. The context would not contain any diagnostics for `ThisMethodExists`. I thought I had seen the lightbulb show incorrectly for this case, but I must be mis-remembering as I cannot reproduce. I'll remove this piece of code. Apologies here!